renku/cli/_providers/zenodo.py

```diff
@@ -0,0 +1,132 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Zenodo API integration."""
+import re
+import urllib
+from urllib.parse import urlparse
+
+import attr
+import requests
+
+from renku.cli._providers.doi import DOIProvider
+
+BASE_URL = 'https://zenodo.org/api/'
+
+
+@attr.s
+class ZenodoFile:
+    """Zenodo record file."""
+
+    checksum = attr.ib()
+    links = attr.ib()
+    bucket = attr.ib()
+    key = attr.ib()
+    size = attr.ib()
+    type = attr.ib()
+
+    @property
+    def remote_url(self):
+        """Get remote URL as ``urllib.ParseResult``."""
+        return urllib.parse.urlparse(self.links['self'])
+
+    @property
+    def name(self):
+        """Get file name."""
+        return self.remote_url.path.split('/')[-1]
+
+
+@attr.s
+class ZenodoRecord:
+    """Zenodo record."""
+
+    id = attr.ib()
+    conceptdoi = attr.ib()
+    conceptrecid = attr.ib()
+
+    doi = attr.ib()
+    files = attr.ib()
+    links = attr.ib()
+    metadata = attr.ib()
+    owners = attr.ib()
+    revision = attr.ib()
+    stats = attr.ib()
+
+    created = attr.ib()
+    updated = attr.ib()
+
+    _zenodo = attr.ib(kw_only=True)
+
+    @property
+    def last_version(self):
+        """Check if record is at last possible version."""
+        return self.version['is_last']
+
+    @property
+    def version(self):
+        """Get record version."""
+        return self.metadata['relations']['version'][0]
+
+    @property
+    def title(self):
+        """Get record title."""
+        return re.sub(r'\W+', '', self.metadata['title']).lower()
+
+    def get_files(self):
+        """Get Zenodo files metadata as ``ZenodoFile``."""
+        if len(self.files) == 0:
+            raise LookupError('no files have been found')
+
+        return [ZenodoFile(**file_) for file_ in self.files]
+
+
+@attr.s
+class ZenodoProvider:
+    """zenodo.org registry API provider."""
+
+    is_doi = attr.ib(default=False)
+
+    @staticmethod
+    def record_id(uri):
+        """Extract record id from uri."""
+        return urlparse(uri).path.split('/')[-1]
+
+    def find_record(self, uri):
+        """Retrieves a record from Zenodo.
+
+        :raises: ``LookupError``
+        :param uri: DOI or URL
+        :return: ``ZenodoRecord``
+        """
+        if self.is_doi:
+            return self.find_record_by_doi(uri)
+
+        return self.get_record(uri)
+
+    def find_record_by_doi(self, doi):
+        """Resolve DOI and make record for retrieved record id."""
+        doi = DOIProvider(uri=doi).as_obj()
+        return self.get_record(ZenodoProvider.record_id(doi.URL))
+
+    def get_record(self, uri):
+        """Retrieve record metadata and return ``ZenodoRecord``."""
+        _url = BASE_URL + 'records/' + ZenodoProvider.record_id(uri)
```

Comment:
joining paths like this makes me nervous... could we use something like

```
import pathlib
import urllib
BASE_URL = 'https://zenodo.org'
BASE_PATH = '/api'
_url = urllib.parse.urljoin(BASE_URL, pathlib.posixpath.join(BASE_PATH, 'records', ZenodoProvider.record_id(uri)))
```

This way you can be sure that all the slashes are correctly taken care of.

Reply:
?
This is URL, not PosixPath => It's same on all platforms. 

Don't be nervous!

Reply:
I know it's not a posixpath but you can use that functionality to join sub-paths together in a predictable way. If someone changes `BASE_URL` and forgets to append the `/` your code will fail with `record not found` and it won't be obvious that the reason is that the URL was mangled.